### DISCUSSION TITLE
Scripting with brew: problems with sudo/root privilege

### DISCUSSION THREAD

**Discussion by i0ntempest** [Homebrew/discussions#2594] [Everyday usage] [answered]
Hi all,

So I have both MacPorts and Homebrew (for casks only) on my computer and I have written a bash script that can upgrade everything installed by them with one command. However I have a few problems with brew and its use of sudo:
1. brew seems to always use /usr/bin/sudo, but I have a custom one installed by MacPorts. Anyway to change brew to use that?
2. brew doesn't want me to execute it with sudo and asks for privilege only if needed. And because of the reason above, it halts my script and ask for password when I already put in my password once for port commands above it. I would like to input my password only once and have the script finish without interruptions (ignoring sudo timeout for now). Anyone have suggestions on how should I do this?

Thanks to all.

**Comment by SMillerDev** [comment, marked as answer]
1. AFAIK that's a deliberate choice to avoid incompatibilities.

2. Check `SUDO_ASKPASS`

**Reply by i0ntempest** [reply]
> * AFAIK that's a deliberate choice to avoid incompatibilities.
> * Check `SUDO_ASKPASS`

Thanks, I did some searching on `SUDO_ASKPASS`. The problem is sometimes I run the update script over ssh and I'm not sure how I should write such a askpass script that ask for password in the same tty and save it for the entire script (without saving the password to disk ofc). I don't think I can cache the password in a variable and call the askpass script with args?

**Reply by carlocab** [reply]
`SUDO_ASKPASS` is just a path to a script that does authentication for you and then tells `sudo` your password if you did that successfully.

The script can be as simple as
```
#!/bin/bash
echo hunter2
```
or
```
#!/bin/bash
echo $MYPASSWORDCACHEDINANENVVARIABLE
```

**Reply by i0ntempest** [reply]
Thanks, I'm now using a heredoc to write a temp script when the script is run. Works great.

**Reply by i0ntempest** [reply]
Actually it isn't working... the shell launched by brew to execute sudo commands not have the password variable I exported in its env so the askpass script echoes back empty. Is writing my password into askpass script the only way?

**Reply by carlocab** [reply]
Prefix the variable with `HOMEBREW_`, so you'll have something like `HOMEBREW_MY_PASSWORD`.
